## Упражнения по библиотеке Numpy

In [2]:
import numpy as np


**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

**1st approach:** by means of mask 

In [14]:
arr = np.arange(10)
arr[(arr > 3) & (arr < 8)] *= -1
arr

array([ 0,  1,  2,  3, -4, -5, -6, -7,  8,  9])

**2nd approach:** by means of np.where method

In [17]:
arr = np.arange(10)
np.where((arr > 3) & (arr < 8), arr*(-1), arr)


array([ 0,  1,  2,  3, -4, -5, -6, -7,  8,  9])

**2.** Заменить максимальный элемент случайного массива на 0

In [16]:
arr = np.arange(10)
arr[np.max(arr)] = 0
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 0])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

**1st approach:** by means of product in intertools modul

In [5]:
import itertools as it
arr1 = np.array([[1,2],[3,4]])
arr2 = np.array([[5,6,7],[8,9,10]])
cartesian = it.product(arr1, arr2)
for i in cartesian:
    print(i)

(array([1, 2]), array([5, 6, 7]))
(array([1, 2]), array([ 8,  9, 10]))
(array([3, 4]), array([5, 6, 7]))
(array([3, 4]), array([ 8,  9, 10]))


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [43]:
np.random.seed(0)
A = np.random.randint(0, 2, size=(8, 3))
B = np.random.randint(0, 2, size=(2, 2))
C = np.unique(B)
mask = [set(C).issubset(set(element)) for element in A]
A[mask]   

array([[0, 1, 1],
       [0, 1, 1],
       [1, 1, 0],
       [0, 1, 0],
       [0, 1, 0],
       [1, 1, 0]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [39]:
np.random.seed(0)
arr = np.random.randint(0, 3, size=(10, 3))
mask = [True if len(np.unique(np.array(element)))>=2 else False for element in arr]
arr[mask]

array([[0, 1, 1],
       [0, 1, 1],
       [1, 1, 0],
       [0, 1, 0],
       [0, 1, 0],
       [1, 1, 0],
       [0, 1, 1],
       [1, 1, 0]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [4]:
arr = np.random.randint(0, 3, size=(8, 3))
np.unique(arr, axis = 0)

array([[0, 2, 2],
       [1, 0, 2],
       [1, 1, 1],
       [1, 2, 2],
       [2, 1, 0]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

**1st approach:** using numpy

In [53]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) 
np.array(np.diagonal(X))[np.array(np.diagonal(X)) != 0].prod()

3

**2nd approach:** without numpy

In [55]:
X = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
prod = 1
for i in range(len(X)):
    for j in range(len(X[i])):
        if i == j and X[i][j] != 0:
            prod *= X[i][j]           

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

**1st approach:** using numpy. In this task we must find out if the numpy arrays consist of the same elements regardless of their order.

In [4]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2]) 
np.array_equal(x.sort(),y.sort())

True

**2nd approach:** without numpy

In [6]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
x.sort() == y.sort()

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

**1st approach:** using numpy

In [13]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
np.max((x[1:][(x==0)[:-1]]))

5

**2nd approach:** without numpy

In [12]:
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
max([x[i] for i in range(1, len(x)) if x[i - 1] == 0])

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

**1st approach:** with numpy

In [5]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
np.unique(x, return_counts=True)

(array([2, 3, 5]), array([3, 3, 1], dtype=int64))

**2nd approach:** without numpy, using dict

In [13]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
result = {}
for element in x:
    result[element] = result.get(element, 0) + 1
values = [key for key in result.keys()]
val_frequency = [values for values in result.values()]
result_coding = (values, val_frequency)
result_coding

([2, 3, 5], [3, 3, 1])

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

**1st approach:** without numpy using scipy.spatial.distance.cdist

In [35]:
%%time
from scipy.spatial import distance
X = [(1,2),(1,3),(1,4)]
Y = [(2,3),(3,4),(4,5)]
distance.cdist(X, Y, 'euclidean')

Wall time: 1.02 ms


array([[1.41421356, 2.82842712, 4.24264069],
       [1.        , 2.23606798, 3.60555128],
       [1.41421356, 2.        , 3.16227766]])

**2nd approach:** with numpy

%%time
X = np.array([(1,2),(1,3),(1,4)])
Y = np.array([(2,3),(3,4),(4,5)])
distance = np.linalg.norm(X-Y, axis = 1)
distance

**Summaries:** numpy provided euclidean distance evaluation much faster than scipy.spatial.distance.cdist

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [39]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [52]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
np.average(calorie_stats)

106.88311688311688

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [43]:
average_calories = np.average(calorie_stats) - 60
average_calories

46.883116883116884

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [49]:
calorie_stats_sorted = np.sort(calorie_stats)
len(calorie_stats_sorted)

77

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [45]:
median_calories = np.median(calorie_stats)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [46]:
nth_percentile = 0
while np.percentile(calorie_stats, nth_percentile) <= 60:
    nth_percentile += 1
nth_percentile

4

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [47]:
more_calories = len(calorie_stats[calorie_stats > 60]) / len(calorie_stats)
more_calories

0.961038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [48]:
calorie_std = np.std(calorie_stats)
calorie_std

19.35718533390827

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

**Summaries:** At the beginning, I will briefly characterize the sample. The sample consists of 77 calorie indicators of cereals of different brands, min = 50, max = 160, avg = 108, median = 110. The distribution of the sample indicators is close to the normal distribution, without abnormal values that could significantly distort the result. The standard error of the mean of 2.2 can be calculated based on the standart deviation, so the confidence interval is: [102.7; 111.3]. Taking the significance level of cliteria = 0.05, we can consider the results of the evaluation to be reliable.

However, it is obvious that for an ordinary consumer, the calculation of indicators of mathematical statistics is not interesting, boring and incomprehensible. The consumer is interested in personal profit, what he will receive and what kind of pain the product will solve for him. Many people carefully monitor weight, it is worth noting for the consumer that CrunchieMunchies cereal is less caloric than 96% of competitors' cereal and the second conclusion is that CrunchieMunchies cereal contains 44% fewer calories than other 96% brands of cereal (the last slogan will be much more attractive and pleasant for the client, because the client immediately sees his benefit: how much calories consumed by him will decrease, consuming CrunchieMunchies instead of other brands).
